### PR TITLE
[BugFix] Fix race condition when calling TransactionState.addPublishVersionTask

### LINKED ISSUE #20277: PublishVersionDaemon.publishVersion got null exception at publishVersionTask.isFinished()
### Steps to reproduce the behavior (Required)
create routine load and run(occasionally)

### Expected behavior (Required)
finished transaction normally and recreate new consumer task

### Real behavior (Required)
1. PublishVersionDaemon throw null exception
2. routine load task stuck in committed state

### StarRocks version (Required)
 - 2.2/2.3/2.4/2.5

### Reason analysis
PublishVersionDaemon and DatabaseTransactionMgr run in parallel, both call TransactionState.addPublishVersionTask method.
It may present with a parallel problem:
1. DatabaseTransactionMgr mark TransactionState to be COMMITTED state
2. PublishVersionDaemon get TransactionState with COMMITTED state
3. PublishVersionDaemon call TransactionState.createPublishVersionTask
4. PublishVersionDaemon submit batchTask to be
5. DatabaseTransactionMgr call transactionState.addPublishVersionTask(backendId, `null`), this with update publishVersionTasks map with null value
6. PublishVersionDaemon call transactionState.getPublishVersionTasks and publishVersionTask.isFinished(), got null exception

### Fix method
1. TransactionState add a set/list field with name `totalInvolvedBackends`
2. DatabaseTransactionMgr add backendid to TransactionState.totalInvolvedBackends
3. TransactionState.createPublishVersionTask implement base on totalInvolvedBackends field

### PR DESCRIPTION
**Pull request by decster**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #20277

## Problem Summary(Required) ：

When committing TXN, OlapTableTxnStateListener calls TransactionState.addPublishVersionTask after txn is set to commit, this will cause a race condition with PublishVersionDaemon thread, it will also call TransactionState.addPublishVersionTask and relies on TransactionState.publishVersionTasks already setup, but it may not be set up. So the fix is to set up TransactionState.publishVersionTasks before set the commit state.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [x] 2.4
  - [x] 2.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/transaction/OlapTableTxnStateListener.java [hunk 1]
```diff
@@ -247,11 +250,6 @@ public void preWriteCommitLog(TransactionState txnState) {
 
     @Override
     public void postWriteCommitLog(TransactionState txnState) {
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/transaction/OlapTableTxnStateListener.java:252]: This function seems unuseful.

**Comment by decster** [comment]
close for now

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.4

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.3

**Pull request closed without merging**